intergov/use_cases/process_channel_notifications.py

```diff
@@ -0,0 +1,117 @@
+
+
+class ChannelNotificationUseCase:
+    CHANNEL_API_GET_MESSAGE_ENDPOINT = 'messages'
+    CHANNEL_ID = 'channel_id'
+    NOTIFICATION_PAYLOAD = 'notification_payload'
+    ATTEMPT = 'attempt'
+
+    def __init__(self, channel_notification_repo: ChannelNotificationRepo):
+        self.channel_notification_repo = channel_notification_repo
+
+
+class EnqueueChannelNotificationUseCase(ChannelNotificationUseCase):
+    def execute(self, channel, notification_payload):
+        """Enqueue job for further processing (processor calls execute() method)"""
+        job_payload = {
+            self.CHANNEL_ID: channel['Id'],
+            self.NOTIFICATION_PAYLOAD: notification_payload,
+            self.ATTEMPT: 1
+        }
+        self._post(job_payload)
+
+    def retry(self, job_payload, attempt):
+        job_payload[self.ATTEMPT] = attempt
+        self._post(job_payload, delay=self._get_retry_time(attempt))
+
+    def _post(self, job_payload, delay=0):
+        self.channel_notification_repo.post_job(job_payload, delay_seconds=delay)
+
+    @staticmethod
+    def _get_retry_time(attempt):
+        """exponential back off with jitter"""
+        if attempt == 1:
+            return 0
+        base = 8
+        max_retry = 100
+        delay = min(base * 2 ** attempt, max_retry)
+        jitter = random.uniform(0, delay / 2)
+        return int(delay / 2 + jitter)
+
+
+class ProcessChannelNotificationUseCase(ChannelNotificationUseCase):
+    MAX_ATTEMPTS = 3
+
+    def __init__(self, channel_notification_repo: ChannelNotificationRepo, bc_inbox_repo: BCInboxRepo, routing_table):
+        super().__init__(channel_notification_repo)
+        self.routing_table = routing_table
+        self.enqueue_message_use_case = EnqueueMessageUseCase(bc_inbox_repo)
+
+    def execute(self):
+        """
+        Get message from channel API and post it to message API.
+        Retry in case of failure
+        """
+
+        job = self.channel_notification_repo.get_job()
+        if not job:
+            return
+
+        queue_message_id, job_payload = job
+        attempt = job_payload[self.ATTEMPT]
+
+        try:
+            self.process(job_payload)
+        except ChannelApiFailure as e:
+            logger.info("%s: processing channel notification failure", queue_message_id)
+            logger.exception(e)
+            if attempt < self.MAX_ATTEMPTS:
+                attempt += 1
+                EnqueueChannelNotificationUseCase(self.channel_notification_repo).retry(job_payload, attempt)
+                return
+            raise
+
+    def process(self, job_payload):
+        channel = job_payload[self.CHANNEL_ID]
+        notification_payload = job_payload[self.NOTIFICATION_PAYLOAD]
+        url = self._get_channel_message_endpoint_url(channel, notification_payload['id'])
+        message = self._get_message(url)
+        self.enqueue_message_use_case.execute(message)
+
+    def _get_message(self, url):
+        response = requests.get(url)
```

Comment:
and what about JWT auth we have for the cloud setup? multichannel worker seems to be handling it on per-channel basis, may be it's worth init this HttpApiChannel class like it does to handle it.

Reply:
currently shared-db-channel doesn't support auth. Let's implement it as separate user story as soon as we have clear requirements

Reply:
https://github.com/trustbridge/trustbridge-hamlet/tree/master/devnet this setup uses transparent auth - underlying code is not aware of this layer

https://sharedchannel.services.devnet.trustbridge.io/